### DISCUSSION TITLE
OFFTOPIC: ESP32 ADC reading in micropython

### DISCUSSION THREAD

**Discussion by luke-jr** [arendst/Tasmota#12626] [Q&A] [answered]
I have an ESP32 I'd like to use for several ADC readings, and implement complex logic paths for. MicroPython seems perfect for the job, but seems to have a huge amount of noise in the ADC readings. Tasmota, on the other hand, is pretty much noise-free on the same board (DevKitC).

In my testing with MicroPython, I did find that with WiFi disabled, the noise goes away. But I can't be disconnecting and reassociating to WiFi every second...

Can someone familiar with Tasmota's ESP32 port help advise how it's avoiding noise while keeping WiFi connected? (So that it can be implemented in MicroPython)

Thanks

**Comment by barbudor** [comment, marked as answer]
Hi @luke-jr 
What you are looking for probably this very simplec code:
https://github.com/arendst/Tasmota/blob/5a5843b5ab9be3e1b93f468a939b48c582c36aed/tasmota/xsns_02_analog.ino#L289-L292

**Reply by luke-jr** [reply]
Hmm, but this doesn't appear to do anything to avoid WiFi interference, so I'm guessing I'd have to dig into `AdcRead` somewhere.

I'm assuming since Tasmota's readings are so noise-free that someone intentionally designed it to avoid the noise issue...

**Reply by sfromis** [reply]
Maybe the setup parameters in the analog driver makes a difference, IDK.
https://github.com/arendst/Tasmota/blob/development/tasmota/xsns_02_analog.ino#L248-L254

**Comment by s-hadinger** [comment]
I'm afraid we don't know much about MicroPython implementation here.

Have you considered using Berry instead? It's much more lightweight than MicroPython. The ADC reading is not yet implemented but that's an easy addition if you need it.
